### ISSUE TITLE
Is there a way to parellelize the bulk insert operations

### ISSUE THREAD

**Issue by stallion2076a** [zzzprojects/EntityFramework-Extensions#7]
Currently I was trying the insert in a single core windows server 2012 , would it matter if we were to bump up the cores . is there a way to achieve parallelism for  the batch operations ?

**Comment by zzzprojects** [comment]
Using parallel will rarely improve performance for SQL Server. In fact, most of the time, the performance will be decreased.

Our library is very IO intensive, so even if you execute multiple SqlBulkCopy in parallel, they will all be way slower since the IO will limit them.

For BulkSaveChanges, the parallel is almost impossible since this operation is a whole transaction and using parallel may cause too many deadlock scenario.

For BulkInsert, BulkUpdate, BulkDelete, and BulkMerge, you can however easily do it if you want to try it: https://github.com/zzzprojects/Bulk-Operations/wiki/Performance-Tuning-Options#parallel

```csharp
context.BulkInsert(new List<EntitiesContext>(), operation =>
{
    operation.UseParallel = true;
    operation.ParallelMinRecord = 100;
});
```

As we said, we don't recommand to enable it for SQL Server. You will rarely get improvement.

**Comment by stallion2076a** [comment]
I am using a bulkinsert followed by a bulk update of 500, 000 records.
I am using batching of 5000 with a batch timeout of 500 . What other performance tuning options do i Have .

thanks

**Comment by zzzprojects** [comment]
By default, BulkInsert returns the identity value and set it to the entity which decreases the performance.

Do your table contains an identity column which you doesn't need to output the value?

It seems no option is available to turn off auto "output". Let us know if this is your situation and we will add this option ASAP.

**Comment by stallion2076a** [comment]
We dont need to output any value just insert the data . I think what you are refferring to should improve performance

**Comment by zzzprojects** [comment]
8 minutes to insert 500,000 records is not usual even when using our library. Do you only benchmark the BulkInsert or you also include some other operations that could heavily impact also this performance? (By example, retrieving data from database).

We just released a new version (3.9.40) and added a new property to disable the Auto Map Output Direction (AutoMapOutputDirection).

```
context.BulkInsert(list, operation =>
{
    operation.AutoMapOutputDirection = false;
    operation.BatchSize = 10000;
    operation.SqlBulkCopyOptions = SqlBulkCopyOptions.FireTriggers | SqlBulkCopyOptions.CheckConstraints | SqlBulkCopyOptions.TableLock;
});
```

**AutoMapOutputDirection**:  By disabling this options, a SqlBulkCopy can be directly be done in the destination table to increase performance.
**BatchSize**: You can decrease/increase the value to find the most suited value for your scenario. The ideal batch size depends on trigger, index, row size, and more, so the only way to find the optimal is by trying some value
**SqlBulkCopyOptions**: By using TableLock as options, you will lock the table but also increase inserts performance by around 50%, enable this option is optional.

Please let me know about your new result.

**Comment by stallion2076a** [comment]
Hello , I have started experiencing an error from this dll after the new version this was not happenning before - Happens during bulk insert
An unhandled exception of type 'System.OutOfMemoryException' occurred in Z.EntityFramework.Extensions.dll

**Comment by zzzprojects** [comment]
How many record did you try to insert?

I believe this issue is due because the bulk insert creates a DataTable from entities to insert.

You can fix this issue by specifying a **DataTableBatchSize** so less memory will be used.
```
using (var ctx = new CodeFirstEntities())
{
    // Examples
    ctx.BulkInsert(list, operation =>
    {
        operation.DataTableBatchSize = 100000;
        operation.BatchSize = 10000;

    });
}
```

We will work on this issue next week to make sure this never happen again.

**Comment by stallion2076a** [comment]
Does AutoMapOutputDirection improve update transactions ? Inserts are performing much better with the change , less than 2 mins .
I have 2 tasks one inserts 500 k records in a destination table 
the other updates the status of 500 k records in the primary table

thanks

**Comment by zzzprojects** [comment]
AutoMapOutputDirection has no impact currently for update transaction.

If you only need to update the status column for 500k records, we recommend you to specify yourself the mapping to allow to update only this column and not all columns.

```csharp
context.BulkUpdate(list, operation =>
{
    operation.ColumnPrimaryKeyExpression = entity1 => entity1.ID;
    operation.ColumnInputExpression = entity1 => entity1.Status;
});
```

**Comment by zzzprojects** [comment]
Let me know about your update performance. There is maybe some optimization which can be done.

**Comment by stallion2076a** [comment]
I think this is working a bit better , can I use 2 columns specified ?

**Comment by zzzprojects** [comment]
Yes, you can map multiple columns using anonymous types

```csharp
context.BulkUpdate(list, operation =>
{
    operation.ColumnPrimaryKeyExpression = entity1 => entity1.ID;
    operation.ColumnInputExpression = entity1 => new { entity1.Status, entity1.Property2 };
});
```

**Comment by stallion2076a** [comment]
This was working fine , but now again am getting out of memory expression with version 3.9.45

**Comment by zzzprojects** [comment]
Hello,

This issue is caused because of DataTable high memory usage. At the beginning of next week, we will refactor the code to use a light version of a DataTable (our homemade implementation). 

Some people report 70MB file growing up to 500MB usage when using a DataTable ([DataTable memory huge consumption](http://stackoverflow.com/questions/18448894/datatable-memory-huge-consumption)), so this for sure doesn't help our library when 500k entities are being inserted.

A fix will be very soon released.

**Comment by zzzprojects** [comment]
After some research & test, we found out our data table implementation was taking less memory, but we were still unhappy with the result.

We modified the code to allow to use a DataReader to "stream" record by record from the entities instead to convert it to a DataTable.

Unless your entities already consume over than 90% of the memory available, you should not have any more this issue with the new version. 

We will release this fix probably Sunday night to make it available for Monday morning.

**Comment by zzzprojects** [comment]
We just release the version 3.10.0 in beta. This change was bigger than expected and we believe additional test should be made before marking this change as stable.

You can try it here:
https://www.nuget.org/packages/Z.EntityFramework.Extensions/3.10.0-beta

DataTable is not longer used in this version unless Parallel is enabled.

Let me know if you still have the Memory Issue.

**Comment by zzzprojects** [comment]
Closing: Released a stable version in v3.10.1
